android/vpnClient/src/main/java/org/mozilla/firefox/vpn/qt/PackageManagerHelper.kt

```diff
@@ -0,0 +1,147 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.mozilla.firefox.vpn.qt
+
+import android.Manifest
+import android.annotation.SuppressLint
+import android.content.Context
+import android.content.Intent
+import android.content.pm.ApplicationInfo
+import android.content.pm.PackageInfo
+import android.content.pm.PackageManager
+import android.graphics.Color
+import android.graphics.drawable.ColorDrawable
+import android.graphics.drawable.Drawable
+import android.net.Uri
+import org.json.JSONObject
+import java.util.ArrayList
+import java.util.Arrays
+
+// Gets used by /platforms/android/androidAppListProvider.cpp
+object PackageManagerHelper {
+    const val TAG = "PackageManagerHelper"
+
+    @Suppress("DEPRECATION")
+    @SuppressLint("QueryPermissionsNeeded") // We have QUERY_ALL_PACKAGES permission
+    @JvmStatic
+    fun getAllAppNames(ctx: Context): JsonAppMap {
+        val output = JSONObject()
+        val pm = ctx.packageManager
+        val browsers = getBrowserIDs(pm)
+        val installedPackages = pm.getInstalledPackages(PackageManager.GET_PERMISSIONS)
+        installedPackages.stream().filter {
+            if (isSelf(it)) {
+                // Cannot exclude ourselves
+                return@filter false
+            }
+            // If it's allowlisted, just keep it
+            if (isAllowListed(it.packageName)) {
+                return@filter true
+            }
+            // Make sure we add all system web-browsers
+            if (browsers.contains(it.packageName)) {
+                return@filter true
+            }
+            // Otherwise just add any app that is not
+            // a system package.
+            return@filter !isSystemPackage(it, pm)
+        }.forEach {
+            output.put(it.packageName, it.applicationInfo.loadLabel(pm).toString())
+        }
+        return output.toString()
+    }
+
+    @JvmStatic
+    fun getAppIcon(ctx: Context, id: ApplicationId): Drawable {
+        try {
+            return ctx.packageManager.getApplicationIcon(id)
+        } catch (e: PackageManager.NameNotFoundException) {
+            e.printStackTrace()
```

Comment:
Can this exception be logged in the VPN logs?

Reply:
i'm afraid not. 
rn we have no access to the vpn-client logstream. 
The daemon has it's own